File: ios/flutter_rasp/Sources/flutter_rasp/FlutterRaspPlugin.swift
import Flutter
import FlutterRaspCore
import UIKit

public class FlutterRaspPlugin: NSObject, FlutterPlugin, FlutterRaspHostApi {

    private let monitoringQueue = DispatchQueue(label: "com.juandpt.flutter_rasp.monitoring", qos: .utility)
    private let screenCaptureManager = ScreenCaptureManager.shared
    private var flutterApi: FlutterRaspFlutterApi?
    private var monitoringTimer: DispatchSourceTimer?
    private let stateLock = NSLock()

    private var _enabledThreats: [String]?
    private var enabledThreats: [String]? {
        get { stateLock.lock(); defer { stateLock.unlock() }; return _enabledThreats }
        set { stateLock.lock(); defer { stateLock.unlock() }; _enabledThreats = newValue }
    }
    private var _exitThreats: [String] = []
    private var exitThreats: [String] {
        get { stateLock.lock(); defer { stateLock.unlock() }; return _exitThreats }
        set { stateLock.lock(); defer { stateLock.unlock() }; _exitThreats = newValue }
    }
    private var monitoringInterval: TimeInterval = 10.0
    private var _isMonitoringActive: Bool = false
    private var isMonitoringActive: Bool {
        get { stateLock.lock(); defer { stateLock.unlock() }; return _isMonitoringActive }
        set { stateLock.lock(); defer { stateLock.unlock() }; _isMonitoringActive = newValue }
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FlutterRaspPlugin()
        FlutterRaspHostApiSetup.setUp(
            binaryMessenger: registrar.messenger(),
            api: instance
        )
        instance.flutterApi = FlutterRaspFlutterApi(
            binaryMessenger: registrar.messenger()
        )
    }


    func startMonitoring(config: RaspConfigMessage, completion: @escaping (Result<Void, Error>) -> Void) {
        enabledThreats = config.enabledThreats
        exitThreats = config.exitThreats
        let interval = config.monitoringIntervalMs

        if let ios = config.iosConfig {
            applyIosConfig(ios)
        }
        Reporter.shared?.setActivePolicy(exitThreats: config.exitThreats)
        DetectorRegistry.shared.clearCache()

        let immediateThreats = DetectorRegistry.shared.detectThreats(enabledThreats: enabledThreats)
        let currentExitThreats = exitThreats
        if !immediateThreats.isEmpty && !currentExitThreats.isEmpty &&
            immediateThreats.contains(where: { currentExitThreats.contains($0) }) {
            let matched = immediateThreats.filter { currentExitThreats.contains($0) }
            logTermination(detected: immediateThreats, matched: matched)
            completion(.success(()))
            monitoringQueue.async { [weak self] in
                self?.terminateApp(matchedExitThreats: matched)
            }
            return
        }

        if !immediateThreats.isEmpty {
            let reportable = immediateThreats.filter { !currentExitThreats.contains($0) }
            if !reportable.isEmpty {
                Reporter.shared?.reportThreatDetected(reportable)
                DispatchQueue.main.async { [weak self] in
                    self?.flutterApi?.onThreatsDetected(threats: reportable) { _ in }
                }
            }
        }

        cancelTimer()
        monitoringInterval = TimeInterval(interval) / 1000.0
        isMonitoringActive = true
        startMonitoringTimer()
        monitoringQueue.async { [weak self] in
            self?.performMonitoringScan()
        }
        addLifecycleObservers()
        completion(.success(()))
    }

    func stopMonitoring(completion: @escaping (Result<Void, Error>) -> Void) {
        isMonitoringActive = false
        cancelTimer()
        enabledThreats = nil
        exitThreats = []
        removeLifecycleObservers()
        completion(.success(()))
    }

    func checkThreat(threatName: String, completion: @escaping (Result<Bool, Error>) -> Void) {
        monitoringQueue.async {
            let detected = DetectorRegistry.shared.detect(threatName: threatName)
            DispatchQueue.main.async {
                completion(.success(detected))
            }
        }
    }

    func scanAll(enabledThreats: [String], completion: @escaping (Result<ScanResultMessage, Error>) -> Void) {
        let threats = enabledThreats.isEmpty ? nil : enabledThreats
        monitoringQueue.async {
            let detected = DetectorRegistry.shared.detectAll(enabledThreats: threats)
            let entries = detected.map { (name, value) in
                ThreatResultEntry(threatName: name, detected: value)
            }
            let message = ScanResultMessage(results: entries)
            DispatchQueue.main.async {
                completion(.success(message))
            }
        }
    }

    func blockScreenCapture(enabled: Bool, completion: @escaping (Result<Void, Error>) -> Void) {
        screenCaptureManager.block(enabled)
        completion(.success(()))
    }

    func isScreenCaptureBlocked(completion: @escaping (Result<Bool, Error>) -> Void) {
        completion(.success(screenCaptureManager.getIsBlocked()))
    }


    func initReporter(
        config: ReporterConfigMessage,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        let coreConfig = ReporterConfig(
            endpoint: config.endpoint,
            headers: config.headers,
            hmacKey: config.hmacKey,
            pinnedCertPem: config.pinnedCertPem.flatMap { Data($0.data) },
            exitTimeoutMs: max(0, Int(config.exitTimeoutMs)),
            httpTimeoutMs: max(100, Int(config.httpTimeoutMs)),
            maxBreadcrumbs: max(1, Int(config.maxBreadcrumbs)),
            maxPendingReports: max(1, Int(config.maxPendingReports)),
            retryBackoffsMs: config.retryBackoffsMs.map { max(100, Int($0)) },
            captureFlutterErrors: config.captureFlutterErrors,
            capturePlatformErrors: config.capturePlatformErrors,
            captureExitThreats: config.captureExitThreats,
            captureDetectedThreats: config.captureDetectedThreats,
            userId: config.userId
        )
        Reporter.initShared(config: coreConfig)
        completion(.success(()))
    }

    func disposeReporter(completion: @escaping (Result<Void, Error>) -> Void) {
        Reporter.disposeShared()
        completion(.success(()))
    }

    func addBreadcrumb(
        breadcrumb: BreadcrumbMessage,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        let data = decodeJsonObject(breadcrumb.dataJson)
        Reporter.shared?.addBreadcrumb(
            category: breadcrumb.category,
            level: breadcrumb.level,
            message: breadcrumb.message,
            data: data,
            timestampMs: breadcrumb.timestampMs
        )
        completion(.success(()))
    }

    func captureError(
        error: CaptureErrorMessage,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        Reporter.shared?.captureException(
            event: error.event,
            message: error.message,
            stackTrace: error.stackTrace,
            library: error.library
        )
        completion(.success(()))
    }

    func setReporterUserId(userId: String?, completion: @escaping (Result<Void, Error>) -> Void) {
        Reporter.shared?.setUserId(userId)
        completion(.success(()))
    }

    func flushReporter(completion: @escaping (Result<Void, Error>) -> Void) {
        Reporter.shared?.flushPending()
        completion(.success(()))
    }


    private func applyIosConfig(_ config: IosConfigMessage) {
        DetectorRegistry.shared.configure(teamId: config.teamId, bundleIds: config.bundleIds)
    }

    private func logTermination(detected: [String], matched: [String]) {
        let separator = String(repeating: "!", count: 60)
        NSLog("[FlutterRASP] %@", separator)
        NSLog("[FlutterRASP]   FLUTTER RASP — SECURITY VIOLATION DETECTED")
        NSLog("[FlutterRASP] %@", separator)
        NSLog("[FlutterRASP]   Detected threats : %@", detected.map { $0.uppercased() }.joined(separator: ", "))
        NSLog("[FlutterRASP]   Policy violation : %@", matched.map { $0.uppercased() }.joined(separator: ", "))
        NSLog("[FlutterRASP]   Action           : TERMINATING APP")
        NSLog("[FlutterRASP] %@", separator)
    }

    private func terminateApp(matchedExitThreats: [String]) {
        // Synchronous; Reporter blocks for at most config.exitTimeoutMs.
        Reporter.shared?.reportExitThreat(matched: matchedExitThreats)
        exit(1)
    }

    private func performMonitoringScan() {
        guard isMonitoringActive else { return }

        let threats = DetectorRegistry.shared.detectThreats(enabledThreats: enabledThreats)
        if !threats.isEmpty {
            let currentExitThreats = exitThreats
            if !currentExitThreats.isEmpty && threats.contains(where: { currentExitThreats.contains($0) }) {
                let matched = threats.filter { currentExitThreats.contains($0) }
                logTermination(detected: threats, matched: matched)
                terminateApp(matchedExitThreats: matched)
                return
            }
            let reportable = threats.filter { !currentExitThreats.contains($0) }
            if !reportable.isEmpty {
                Reporter.shared?.reportThreatDetected(reportable)
                DispatchQueue.main.async { [weak self] in
                    self?.flutterApi?.onThreatsDetected(threats: reportable) { _ in }
                }
            }
        }
    }

    private func startMonitoringTimer() {
        let timer = DispatchSource.makeTimerSource(queue: monitoringQueue)
        timer.schedule(
            deadline: .now() + monitoringInterval,
            repeating: monitoringInterval,
            leeway: .milliseconds(100)
        )
        timer.setEventHandler { [weak self] in
            self?.performMonitoringScan()
        }
        monitoringTimer = timer
        timer.resume()
    }

    private func cancelTimer() {
        monitoringTimer?.cancel()
        monitoringTimer = nil
    }

    private func addLifecycleObservers() {
        removeLifecycleObservers()
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appDidBecomeActive),
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appWillResignActive),
            name: UIApplication.willResignActiveNotification,
            object: nil
        )
    }

    private func removeLifecycleObservers() {
        NotificationCenter.default.removeObserver(self, name: UIApplication.didBecomeActiveNotification, object: nil)
        NotificationCenter.default.removeObserver(self, name: UIApplication.willResignActiveNotification, object: nil)
    }

    @objc private func appDidBecomeActive() {
        guard isMonitoringActive else { return }
        DetectorRegistry.shared.clearCache()
        cancelTimer()
        startMonitoringTimer()
        monitoringQueue.async { [weak self] in
            self?.performMonitoringScan()
        }
    }

    @objc private func appWillResignActive() {
        cancelTimer()
    }

    private func decodeJsonObject(_ json: String) -> [String: Any] {
        guard !json.isEmpty,
              let data = json.data(using: .utf8),
              let dict = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        else { return [:] }
        return dict
    }
}

File: android/src/main/kotlin/com/juandpt/flutter_rasp/FlutterRaspPlugin.kt
package com.juandpt.flutter_rasp

import android.app.Activity
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.juandpt.flutter_rasp_core.DetectorRegistry
import com.juandpt.flutter_rasp_core.ScreenCaptureManager
import com.juandpt.flutter_rasp_core.reporter.Reporter
import com.juandpt.flutter_rasp_core.reporter.ReporterConfig
import com.juandpt.flutter_rasp_core.reporter.collectors.AppMetadataCollector
import com.juandpt.flutter_rasp_core.reporter.collectors.DeviceIdProvider
import com.juandpt.flutter_rasp_core.reporter.collectors.DeviceMetadataCollector
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

class FlutterRaspPlugin : FlutterPlugin, FlutterRaspHostApi, ActivityAware {

    companion object {
        private const val TAG = "FlutterRASP"
    }

    private var flutterApi: FlutterRaspFlutterApi? = null
    private var applicationContext: android.content.Context? = null
    @Volatile private var activity: Activity? = null
    private val screenCaptureManager = ScreenCaptureManager()
    private val mainHandler = Handler(Looper.getMainLooper())
    private var scheduledExecutor: ScheduledExecutorService? = null
    private var monitoringFuture: ScheduledFuture<*>? = null
    @Volatile private var enabledThreats: List<String>? = null
    @Volatile private var exitThreats: List<String> = emptyList()
    @Volatile private var monitoringActive = false

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        FlutterRaspHostApi.setUp(binding.binaryMessenger, this)
        flutterApi = FlutterRaspFlutterApi(binding.binaryMessenger)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        FlutterRaspHostApi.setUp(binding.binaryMessenger, null)
        flutterApi = null
        stopMonitoringInternal()
        enabledThreats = null
        exitThreats = emptyList()
        applicationContext = null
    }


    override fun startMonitoring(config: RaspConfigMessage, callback: (Result<Unit>) -> Unit) {
        val context = applicationContext
        if (context == null) {
            callback(noContextFailure())
            return
        }

        enabledThreats = config.enabledThreats
        exitThreats = config.exitThreats
        val interval = config.monitoringIntervalMs.toLong()
        config.androidConfig?.let { applyAndroidConfig(it) }
        Reporter.get()?.setActivePolicy(config.exitThreats)

        startMonitoringInternal(context, interval)
        callback(Result.success(Unit))
    }

    override fun stopMonitoring(callback: (Result<Unit>) -> Unit) {
        stopMonitoringInternal()
        enabledThreats = null
        exitThreats = emptyList()
        callback(Result.success(Unit))
    }

    override fun checkThreat(threatName: String, callback: (Result<Boolean>) -> Unit) {
        val context = applicationContext
        if (context == null) {
            callback(noContextFailure())
            return
        }
        Thread {
            val detected = DetectorRegistry.detect(threatName, context)
            mainHandler.post { callback(Result.success(detected)) }
        }.start()
    }

    override fun scanAll(enabledThreats: List<String>, callback: (Result<ScanResultMessage>) -> Unit) {
        val context = applicationContext
        if (context == null) {
            callback(noContextFailure())
            return
        }
        Thread {
            val detected = DetectorRegistry.detectAll(context, enabledThreats)
            val entries = detected.map { (name, value) ->
                ThreatResultEntry(threatName = name, detected = value)
            }
            val message = ScanResultMessage(results = entries)
            mainHandler.post { callback(Result.success(message)) }
        }.start()
    }

    override fun blockScreenCapture(enabled: Boolean, callback: (Result<Unit>) -> Unit) {
        val currentActivity = activity
        if (currentActivity == null) {
            callback(
                Result.failure(
                    FlutterError(
                        "SCREEN_CAPTURE_NO_ACTIVITY",
                        "No active Activity available to block screen capture",
                        null,
                    ),
                ),
            )
            return
        }
        screenCaptureManager.block(currentActivity, enabled)
        callback(Result.success(Unit))
    }

    override fun isScreenCaptureBlocked(callback: (Result<Boolean>) -> Unit) {
        callback(Result.success(screenCaptureManager.isBlocked()))
    }


    override fun initReporter(
        config: ReporterConfigMessage,
        callback: (Result<Unit>) -> Unit,
    ) {
        val context = applicationContext
        if (context == null) {
            callback(noContextFailure())
            return
        }
        try {
            Reporter.init(context, config.toCoreConfig())
            callback(Result.success(Unit))
        } catch (e: Throwable) {
            Log.w(TAG, "initReporter failed: ${e.message}")
            callback(Result.failure(e))
        }
    }

    override fun disposeReporter(callback: (Result<Unit>) -> Unit) {
        Reporter.dispose()
        callback(Result.success(Unit))
    }

    override fun addBreadcrumb(
        breadcrumb: BreadcrumbMessage,
        callback: (Result<Unit>) -> Unit,
    ) {
        Reporter.get()?.addBreadcrumb(
            category = breadcrumb.category,
            level = breadcrumb.level,
            message = breadcrumb.message,
            data = decodeJsonObject(breadcrumb.dataJson),
            timestampMs = breadcrumb.timestampMs,
        )
        callback(Result.success(Unit))
    }

    override fun captureError(
        error: CaptureErrorMessage,
        callback: (Result<Unit>) -> Unit,
    ) {
        Reporter.get()?.captureException(
            event = error.event,
            message = error.message,
            stackTrace = error.stackTrace,
            library = error.library,
        )
        callback(Result.success(Unit))
    }

    override fun setReporterUserId(userId: String?, callback: (Result<Unit>) -> Unit) {
        Reporter.get()?.setUserId(userId)
        callback(Result.success(Unit))
    }

    override fun flushReporter(callback: (Result<Unit>) -> Unit) {
        Reporter.get()?.flushPending()
        callback(Result.success(Unit))
    }


    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }


    private fun applyAndroidConfig(config: AndroidConfigMessage) {
        DetectorRegistry.configure(config.signingCertHashes, config.supportedStores)
    }

    private fun startMonitoringInternal(context: android.content.Context, intervalMs: Long) {
        stopMonitoringInternal()
        monitoringActive = true
        val executor = Executors.newSingleThreadScheduledExecutor { runnable ->
            Thread(runnable, "flutter-rasp-monitor").apply {
                isDaemon = true
                priority = Thread.NORM_PRIORITY - 1
            }
        }
        scheduledExecutor = executor
        monitoringFuture = executor.scheduleAtFixedRate({
            try {
                if (!monitoringActive) return@scheduleAtFixedRate
                val currentEnabled = enabledThreats
                val currentExitThreats = exitThreats
                val threats = DetectorRegistry.detectThreats(context, currentEnabled)
                if (threats.isNotEmpty()) {
                    if (currentExitThreats.isNotEmpty() && threats.any { it in currentExitThreats }) {
                        val matched = threats.filter { it in currentExitThreats }
                        logTermination(threats, matched)
                        terminateApp(matched)
                        return@scheduleAtFixedRate
                    }
                    val reportable = threats.filter { it !in currentExitThreats }
                    if (reportable.isNotEmpty()) {
                        Reporter.get()?.reportThreatDetected(reportable)
                        mainHandler.post {
                            flutterApi?.onThreatsDetected(reportable) {}
                        }
                    }
                }
            } catch (_: Throwable) {
                // Throwable: anything escaping cancels all future runs.
            }
        }, 0L, intervalMs, TimeUnit.MILLISECONDS)
    }

    private fun logTermination(detected: List<String>, matched: List<String>) {
        val separator = "!".repeat(60)
        Log.e(TAG, separator)
        Log.e(TAG, "  FLUTTER RASP — SECURITY VIOLATION DETECTED")
        Log.e(TAG, separator)
        Log.e(TAG, "  Detected threats : ${detected.joinToString(", ") { it.uppercase() }}")
        Log.e(TAG, "  Policy violation : ${matched.joinToString(", ") { it.uppercase() }}")
        Log.e(TAG, "  Action           : TERMINATING APP")
        Log.e(TAG, separator)
    }

    private fun terminateApp(matchedExitThreats: List<String>) {
        // Synchronous; the reporter caps itself at config.exitTimeoutMs.
        runCatching { Reporter.get()?.reportExitThreat(matchedExitThreats) }

        val currentActivity = activity
        if (currentActivity != null) {
            mainHandler.post {
                try {
                    currentActivity.finishAndRemoveTask()
                } catch (_: Exception) {
                    try { currentActivity.finishAffinity() } catch (_: Exception) {}
                }
                android.os.Process.killProcess(android.os.Process.myPid())
            }
        }

        Thread {
            Thread.sleep(500)
            Runtime.getRuntime().exit(1)
        }.start()
    }

    private fun stopMonitoringInternal() {
        monitoringActive = false
        monitoringFuture?.cancel(false)
        monitoringFuture = null
        scheduledExecutor?.shutdown()
        scheduledExecutor = null
    }

    private fun noContextFailure(): Result<Nothing> = Result.failure(
        FlutterError("NO_CONTEXT", "Application context is not available", null),
    )

    private fun decodeJsonObject(json: String): Map<String, Any?> {
        if (json.isEmpty()) return emptyMap()
        return runCatching {
            val obj = org.json.JSONObject(json)
            val map = mutableMapOf<String, Any?>()
            val keys = obj.keys()
            while (keys.hasNext()) {
                val k = keys.next()
                map[k] = obj.opt(k).takeUnless { it == org.json.JSONObject.NULL }
            }
            map.toMap()
        }.getOrDefault(emptyMap())
    }

    private fun ReporterConfigMessage.toCoreConfig() = ReporterConfig(
        endpoint = endpoint,
        headers = headers,
        hmacKey = hmacKey,
        pinnedCertPem = pinnedCertPem,
        exitTimeoutMs = exitTimeoutMs.coerceAtLeast(0L),
        httpTimeoutMs = httpTimeoutMs.toInt().coerceAtLeast(100),
        maxBreadcrumbs = maxBreadcrumbs.toInt().coerceAtLeast(1),
        maxPendingReports = maxPendingReports.toInt().coerceAtLeast(1),
        retryBackoffsMs = retryBackoffsMs.map { it.coerceAtLeast(100L) },
        captureFlutterErrors = captureFlutterErrors,
        capturePlatformErrors = capturePlatformErrors,
        captureExitThreats = captureExitThreats,
        captureDetectedThreats = captureDetectedThreats,
        userId = userId,
    )
}
